docs/proposals/selinux.md

```diff
@@ -0,0 +1,261 @@
+distributed file systems support `chown`, `chmod`, and `chcon` operations uniformly.  Our API should
+make it simple to express whether a particular volume should have these concerns managed by
+Kubernetes.  For simplicity's sake, we will assume that if a volume requires ownership management,
+that it also requires relabeling and vice-versa.
+
+#### API requirements
+
+From the above, we know that label management must be applied:
+
+1.  To some volume types always
+2.  To some volume types never
+3.  To some volume types *sometimes*
+
+If the SELinux cluster capability is enabled, volumes should be relabeled with the correct SELinux
+context.  Docker has this capability today; it is desireable for other container runtime
+implementations to provide similar functionality.
+
+Relabeling should be an optional aspect of a volume plugin to accomodate:
+
+1.  volume types for which generalized relabeling support is not sufficient
+2.  testing for each volume plugin individually
+
+## Proposed Design
+
+Our design should minimize code for handling SELinux labelling required in the Kubelet and volume
+plugins.
+
+### Deferral: MCS label allocation
+
+Our short-term goal is to facilitate volume sharing and isolation with SELinux and expose the
+primitives for higher level composition; making these automatic is a longer-term goal.  Allocating
+groups and MCS labels are fairly complex problems in their own right, and so our proposal will not
+encompass either of these topics.  There are several problems that the solution for allocation
+depends on:
+
+1.  Users and groups in Kubernetes
+2.  General auth policy in Kubernetes
+3.  [security policy](https://github.com/GoogleCloudPlatform/kubernetes/pull/7893)
+
+### API changes
+
+We will add a new field to the `PodSecurityContext` to capture the pod-level SELinux context:
+
+```go
+package api
+
+type PodSecurityContext struct {
+	// SELinuxOptions captures the SELinux context for all containers in a Pod
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"
+}
+```
+
+The V1 API will be extended with the same field:
+
+```go
+package v1
+
+type PodSecurityContext struct {
+	// SELinuxOptions captures the SELinux context for all containers in a Pod
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"
+}
+```
+
+#### API backward compatibility
+
+TODO
+
+### Kubelet changes
+
+The Kubelet should be modified to perform label management when required for a volume.  The criteria
+to activate the kubelet label management for volumes are:
+
+1.  SELinux integration is enabled in the cluster
+2.  SELinux is enabled on the node
+3.  The pod-level SELinux context is populated
+4.  The volume plugin supports label management
+
+The `volume.Builder` interface should have a new method added that indicates whether the plugin
+supports label management:
+
+```go
+package volume
+
+type Builder interface {
+    // other methods omitted
+    SupportsLabelManagement() bool
```

Comment:
What does it mean to support label management?  What does a plugin have to DO to "earn" this badge?

And can we prettyu please give it a better name.  "label" has a pretty well-known meaning in kubernetes... SecLabel, SELinuxLabel, somethgn...